gh-128150: improve performances of `uuid.uuid*` constructor functions.

picnixz opened the pull request:
There are some points that can be addressed:

- We can drop some micro-optimizations to reduce the diff. Most of the time is taken by function calls and loading integers.
- ~~HACL* MD5 *is* faster than OpenSSL MD5 so it's better to use the former. However, using `_md5.md5` or `from _md5 import md5` is a micro-optimization that can be dropped without affecting performances too much (see https://github.com/python/cpython/pull/128151/commits/cff86e9e8d5318a2f933fafdbf191131c5ebf627 and https://github.com/python/cpython/pull/128151/commits/7095aa4fa92a8e449a0de7a7503fac4bd7be0d60)~~ For consistency, we'll rely on OpenSSL-based implementation even if it's a bit slower.
- ~~The rationale of expanding `not 0 <= x < 1 << 128` into `x < 0 or x > 0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff` is due to the non-equivalent bytecodes. 
  Similar arguments apply to expanding `not 0 <= x < (1 << C)` into `x < 0 or x > B` where B is the hardcoded hexadecimal value of `(1 << C) - 1`.~~


  <details>
  <summary>Bytecode comparisons (not useful, constant folding will make them similarly performant)</summary>

	```text
	  1           LOAD_SMALL_INT           0
	              LOAD_NAME                0 (x)
	              SWAP                     2
	              COPY                     2
	              COMPARE_OP              42 (<=)
	              COPY                     1
	              TO_BOOL
	              POP_JUMP_IF_FALSE        9 (to L1)
	              NOT_TAKEN
	              POP_TOP
	              LOAD_SMALL_INT           1
	              LOAD_SMALL_INT         128
	              BINARY_OP                3 (<<)
	              COMPARE_OP               2 (<)
	              JUMP_FORWARD             2 (to L2)
	      L1:     SWAP                     2
	              POP_TOP
	      L2:     TO_BOOL
	              UNARY_NOT
	              POP_TOP
	              LOAD_CONST               0 (None)
	              RETURN_VALUE
	```
	
	versus
	
	```text
	  1           LOAD_NAME                0 (x)
	              LOAD_SMALL_INT           0
	              COMPARE_OP               2 (<)
	              COPY                     1
	              TO_BOOL
	              POP_JUMP_IF_TRUE         8 (to L1)
	              POP_TOP
	              LOAD_NAME                0 (x)
	              LOAD_CONST               0 (340282366920938463463374607431768211455)
	              COMPARE_OP             132 (>)
	              POP_TOP
	              LOAD_CONST               1 (None)
	              RETURN_VALUE
	      L1:     POP_TOP
	              LOAD_CONST               1 (None)
	              RETURN_VALUE
	```

</details>

<!-- gh-issue-number: gh-128150 -->
* Issue: gh-128150
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--128151.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

picnixz committed:
improve performance of UUIDs creation

picnixz committed:
add What's New entry

picnixz committed:
blurb

picnixz committed:
fix issue number

picnixz committed:
fix typos

picnixz committed:
ensure 14-bit clock sequence

picnixz committed:
Merge branch 'main' into perf/uuid/init-128150

picnixz committed:
add dedicated private fast constructor

picnixz committed:
revert UUIDv1 construction

picnixz committed:
change eager check into an assertion check for internal constructor

picnixz committed:
update performance results

eendebakpt commented on Lib/uuid.py:
```diff
@@ -225,6 +237,15 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
         object.__setattr__(self, 'is_safe', is_safe)
 
+    @classmethod
+    def _from_int(cls, int, *, is_safe=SafeUUID.unknown):
```
```suggestion
    def _from_int(cls, int):
```

eendebakpt commented on Lib/uuid.py:
```diff
@@ -225,6 +237,15 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
+        assert int >= 0 and int <= 0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff
+        self = cls.__new__(cls)
+        object.__setattr__(self, 'int', int)
+        object.__setattr__(self, 'is_safe', is_safe)
```
```suggestion
```
At this moment the argument is unused. Removing it makes the call faster.

picnixz replied:
We still need to create the property, but I can hardcode the value.

eendebakpt commented on Lib/uuid.py:
```diff
@@ -700,24 +721,35 @@ def uuid3(namespace, name):
-        usedforsecurity=False
-    ).digest()
-    return UUID(bytes=digest[:16], version=3)
+    # HACL*-based MD5 is slightly faster than its OpenSSL version,
```
I would put these comments in the pr comments and leave them out of the code. (but I do see the value of them)

picnixz replied:
Done in the commits actually.

eendebakpt commented on Doc/whatsnew/3.14.rst:
```diff
@@ -668,6 +668,25 @@ io
+  * :func:`~uuid.uuid3` is 47% faster for 16-byte names and 13% faster
+    for 1024-byte names. Performances for longer names remain unchanged.
+  * :func:`~uuid.uuid5` is 35% faster for 16-byte names and 24% faster
+    for 1024-byte names. Performances for longer names remain unchanged.
```
```suggestion
    for 1024-byte names. Performance for longer names remains unchanged.
```

eendebakpt commented on Doc/whatsnew/3.14.rst:
```diff
@@ -668,6 +668,25 @@ io
+  functions:
+
+  * :func:`~uuid.uuid3` is 47% faster for 16-byte names and 13% faster
+    for 1024-byte names. Performances for longer names remain unchanged.
```
```suggestion
    for 1024-byte names. Performance for longer names remains unchanged.
```

eendebakpt replied:
```suggestion
    for 1024-byte names. Performance for longer names remains unchanged.
```

eendebakpt commented on Doc/whatsnew/3.14.rst:
```diff
@@ -668,6 +668,25 @@ io
+    for 1024-byte names. Performances for longer names remain unchanged.
+  * :func:`~uuid.uuid4` is 33% faster and :func:`~uuid.uuid8` is 38% faster.
+
+  Overall, dedicated generation of UUID objects version 3, 4, 5, and 8 is
```
Since you already have the specific improvements I would remove this paragraph (or the other way around)

picnixz replied:
I think it's nice for people who just want an overview (like a TL;DR)

picnixz replied:
Ok, I've removed it since I kept it in the simple changelog.

eendebakpt reviewed (commented):
Nice improvement overall! Personally I am not a fan of the lazy imports here, but I'll let someone else decide on that.

picnixz committed:
describe constants

picnixz committed:
revert UUIDv1 optimizations to reduce the diff

picnixz committed:
simplify `_from_int` private constructor as per Pieter's review

picnixz committed:
revert micro-optimization of `not a <= x <= b`

picnixz committed:
use built-in `int` when it is not shadowed

picnixz committed:
remove rationale comment for HACL* MD5

In this commit, we move the rationale for using HACL*-based MD5
instead of its OpenSSL implementation from the code to this note.

HACL*-based MD5 is 2x faster than its OpenSSL implementation for
creating the hash object via `h = md5(..., usedforsecurity=False)`
but `h.digest()` is slightly (yet noticeably) slower.

Overall, HACL*-based MD5 still remains faster than its OpenSSL-based
implementation, whence the choice of `_md5.md5` over `hashlib.md5`.

picnixz committed:
remove rationale comment for OpenSSL SHA-1

In this commit, we move the rationale for using OpenSSL-based SHA-1
instead of its HACL* implementation from the code to this note.

HACL*-based SHA-1 is 2x faster than its OpenSSL implementation for
creating the hash object via `h = sha1(..., usedforsecurity=False)`
but `h.digest()` is almost 3x slower.

Unlike HACL* MD5, HACL*-based SHA-1 is slower than its OpenSSL-based
implementation, whence the choice of `hashlib.sha1` over `_sha1.sha1`.

picnixz committed:
clear variant and version bits using dedicated mask

picnixz committed:
fix typos

picnixz committed:
update benchmarks

eendebakpt commented on Lib/uuid.py:
```diff
@@ -174,57 +189,70 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
+    @classmethod
+    def _from_int(cls, value):
+        """Internal use only."""
+        assert isinstance(value, int), repr(value)
```
The asserts here take a bit of computation time. Since it is for internal use only, I would be fine with removing them

picnixz replied:
I think we can remove this assert but let's leave the range assert.

eendebakpt commented on Lib/uuid.py:
```diff
@@ -700,24 +728,31 @@ def uuid3(namespace, name):
-    return UUID(bytes=digest[:16], version=3)
+    import _md5
+    h = _md5.md5(namespace.bytes + name, usedforsecurity=False)
+    assert len(h.digest()) == 16
```
Also this takes time, in particular the `.digest()`

picnixz replied:
Ah right, I generally use `python -OO` when I'm dealing with performance-critical scripts. Nonethless, I can compute `h.digest()` outside the assertion and check its length afterwards.

picnixz replied:
I've removed the assert altogether. If the MD5 digest is not 16 bytes, something is fundamentally wrong elsewhere.

eendebakpt commented on Lib/uuid.py:
```diff
@@ -740,7 +775,9 @@ def uuid8(a=None, b=None, c=None):
     int_uuid_8 |= c & 0x3fff_ffff_ffff_ffff
-    return UUID(int=int_uuid_8, version=8)
+    # by construction, the variant and version bits are already cleared
+    int_uuid_8 |= _RFC_4122_VERSION_8_FLAGS
```
```suggestion
    int_uuid_8 += _RFC_4122_VERSION_8_FLAGS
```
I think we should not make this change (the `|=` is more readable), but still fun to wonder: would it be faster due to the adaptive interpreter?

picnixz replied:
Well.. in this case it works well but if we were to change the construction, we'll also need to update the operation. I think it's clearer with a `|` rather than a `+` although I said that the bits were cleared. I haven't benchmark your alternative though.

eendebakpt replied:
It makes a tiny difference, but lets indeed stick with `|`
```
In [9]: %timeit int_uuid_8 | _RFC_4122_VERSION_8_FLAGS
   ...: %timeit int_uuid_8 + _RFC_4122_VERSION_8_FLAGS
43.9 ns ± 0.231 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
39.3 ns ± 0.235 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

eendebakpt reviewed (approved):
I left a few small remarks, but am also ok with leaving the code as is. The PR improves performance for the main cases, does not make the other cases smaller and code complexity is roughly the same.

picnixz committed:
remove un-necessary assertions

picnixz committed:
use `object.__new__` instead of `cls.__new__`

hauntsaninja commented on Lib/uuid.py:
```diff
@@ -174,57 +189,69 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
             (time_low, time_mid, time_hi_version,
              clock_seq_hi_variant, clock_seq_low, node) = fields
-            if not 0 <= time_low < 1<<32:
+            if not 0 <= time_low <= 0xffff_ffff:
```
can we leave these alone? lhs reads clearer to me, seems churny, cpython will constant fold so there's no speed improvement

hauntsaninja commented on Lib/uuid.py:
```diff
@@ -85,6 +85,21 @@ class SafeUUID:
+_UINT_128_MAX = 0xffff_ffff_ffff_ffff_ffff_ffff_ffff_ffff
+# 128-bit mask to clear the variant and version bits of a UUID integral value
+#
+# This is equivalent to the 2-complement of '(0xc000 << 48) | (0xf000 << 64)'.
```
we could use the formulae and let cpython constant fold here as well

picnixz committed:
Merge branch 'main' into perf/uuid/init-128150

picnixz committed:
remove dedicated constant folding

picnixz committed:
update benchmarks

encukou commented on Lib/uuid.py:
```diff
@@ -700,24 +723,30 @@ def uuid3(namespace, name):
-        usedforsecurity=False
-    ).digest()
-    return UUID(bytes=digest[:16], version=3)
+    import _md5
```
Could you go through `hashlib` here? `_md5` is a *fallback* implementation; the `usedforsecurity` argument is useless if you use `_md5` directly.

picnixz replied:
Using OpenSSL-based MD5 implementation is a bit slower. Since we're ~~always~~
bundling HACL* MD5 implementation, I wondered whether we could just use it. Or do you think users would prefer if we use `hashlib` explicitly (and OpenSSL when available)?

While `usedforsecurity` is entirely ignored, it wouldn't hurt keeping the semantic meaning (and if we're planning to provide a meaning to `usedforsecurity` for HACL*-based hashes in the future).

But if I can revert that part (we're still going faster I think but the improvements are less pronounced; I'll just need to recompute the numbers (tomorrow))

picnixz replied:
After thinking a bit more, let's just revert to using `hashlib`. We don't lose too much and it would be more consistent with `uuid5`.

picnixz committed:
Always use `hashlib.md5` for consistency as per Petr's comment.

picnixz committed:
update benchmarks

encukou reviewed (approved):
>  Since we're always bundling `HACL*` MD5 implementation, I wondered whether we could just use it. Or do you think users would prefer if we use `hashlib` explicitly (and OpenSSL when available)?

It's possible to [build Python without HACL*](https://docs.python.org/3/using/configure.html#cmdoption-with-builtin-hashlib-hashes), mainly for environments where cryptography is somehow regulated (i.e. the infamous FIPS mode).

It might be OK to use `HACL*` by default *in hashlib*, possibly making it a bit faster for everyone (who didn't opt out).

But UUID should, IMO, use the default. We agree on that, hit the green button :)

picnixz committed:
Merge branch 'main' into perf/uuid/init-128150

picnixz committed:
update NEWS
